Remove the concept of window "key levels".

hlinnaka opened the pull request:
It wasn't very useful. ORCA and Postgres both just stack WindowAgg nodes
on top of each other, and no-one's been unhappy about that, so we might as
well do that, too. This reduces the difference between GPDB and the upstream
implementation, and will hopefully make it smoother to switch.

Rename the Window Plan node type to WindowAgg, to match upstream, now
that it is fairly close to the way the upstream version.

hsyuan commented on src/backend/optimizer/plan/planwindow.c:
```diff
@@ -2295,14 +2279,16 @@ static Plan *plan_sequential_stage(PlannerInfo *root,
+							   winfo->partkey_len, partkey, partkey_operators,
+							   winfo->ordNumCols, winfo->ordColIdx, winfo->ordOperators,
+							   winfo->frameOptions,
+							   /* XXX mutate windowKeys to translate any Var nodes in frame vals. */
```
The comment is outdated, right?

d commented:
I believe this change have resulted in inferior plans. Haven't looked too closely at exactly what's causing it, but here are the steps to reproduce:


1. DDL:

    ```SQL
    CREATE TABLE foo (a int, b int, c int);
    ```

1. Queries:

    ```SQL
    SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);

    SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
    ```

1. Plans after your patch:

    ```SQL
    EXPLAIN SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                              QUERY PLAN
    ----------------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=14705.73..16068.98 rows=77900 width=12)
       ->  Subquery Scan coplan  (cost=14705.73..16068.98 rows=25967 width=12)
             ->  WindowAgg  (cost=14705.73..15289.98 rows=25967 width=12)
                   Partition By: foo.a
                   Order By: foo.b, foo.c
                   ->  Sort  (cost=14705.73..14900.48 rows=25967 width=12)
                         Sort Key: foo.a, foo.b, foo.c
                         ->  WindowAgg  (cost=7208.12..7597.62 rows=25967 width=12)
                               Partition By: foo.a
                               Order By: foo.b
                               ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                                     Sort Key: foo.a, foo.b
                                     ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (15 rows)
    
    EXPLAIN SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                           QUERY PLAN
    -----------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=14705.73..15289.98 rows=77900 width=12)
       ->  WindowAgg  (cost=14705.73..15289.98 rows=25967 width=12)
             Partition By: foo.a
             Order By: foo.b, foo.c
             ->  Sort  (cost=14705.73..14900.48 rows=25967 width=12)
                   Sort Key: foo.a, foo.b, foo.c
                   ->  WindowAgg  (cost=7208.12..7597.62 rows=25967 width=12)
                         Partition By: foo.a
                         Order By: foo.b
                         ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                               Sort Key: foo.a, foo.b
                               ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (14 rows)
    ```

1. Plans before your patch:

    ```SQL
    EXPLAIN SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                          QUERY PLAN
    ---------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=7208.12..7987.12 rows=77900 width=12)
       ->  Window  (cost=7208.12..7987.12 rows=25967 width=12)
             Partition By: a
             Order By (level 1): b
             Order By (level 2): c
             ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                   Sort Key: a, b, c
                   ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (10 rows)
    
    EXPLAIN SELECT *, count(*) over (jz) as jesse, count(*) over (oa) as omer FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                          QUERY PLAN
    ---------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=7208.12..7987.12 rows=77900 width=12)
       ->  Window  (cost=7208.12..7987.12 rows=25967 width=12)
             Partition By: a
             Order By (level 1): b
             Order By (level 2): c
             ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                   Sort Key: a, b, c
                   ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (10 rows)

    ```

Renames aside, we do see an extra sort after your change. And depending on the data this could be quite taxing on performance.

d commented:
I believe this change has resulted in inferior plans. Haven't looked too closely at exactly what's causing it, but here are the steps to reproduce:


1. DDL:

    ```SQL
    CREATE TABLE foo (a int, b int, c int);
    ```

1. Queries:

    ```SQL
    SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);

    SELECT *, count(*) over (jz) as jesse, count(*) over (oa) as omer FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
    ```

1. Plans after your patch:

    ```SQL
    EXPLAIN SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                              QUERY PLAN
    ----------------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=14705.73..16068.98 rows=77900 width=12)
       ->  Subquery Scan coplan  (cost=14705.73..16068.98 rows=25967 width=12)
             ->  WindowAgg  (cost=14705.73..15289.98 rows=25967 width=12)
                   Partition By: foo.a
                   Order By: foo.b, foo.c
                   ->  Sort  (cost=14705.73..14900.48 rows=25967 width=12)
                         Sort Key: foo.a, foo.b, foo.c
                         ->  WindowAgg  (cost=7208.12..7597.62 rows=25967 width=12)
                               Partition By: foo.a
                               Order By: foo.b
                               ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                                     Sort Key: foo.a, foo.b
                                     ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (15 rows)
    
    EXPLAIN SELECT *, count(*) over (jz) as jesse, count(*) over (oa) as omer FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                           QUERY PLAN
    -----------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=14705.73..15289.98 rows=77900 width=12)
       ->  WindowAgg  (cost=14705.73..15289.98 rows=25967 width=12)
             Partition By: foo.a
             Order By: foo.b, foo.c
             ->  Sort  (cost=14705.73..14900.48 rows=25967 width=12)
                   Sort Key: foo.a, foo.b, foo.c
                   ->  WindowAgg  (cost=7208.12..7597.62 rows=25967 width=12)
                         Partition By: foo.a
                         Order By: foo.b
                         ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                               Sort Key: foo.a, foo.b
                               ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (14 rows)
    ```

1. Plans before your patch:

    ```SQL
    EXPLAIN SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                          QUERY PLAN
    ---------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=7208.12..7987.12 rows=77900 width=12)
       ->  Window  (cost=7208.12..7987.12 rows=25967 width=12)
             Partition By: a
             Order By (level 1): b
             Order By (level 2): c
             ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                   Sort Key: a, b, c
                   ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (10 rows)
    
    EXPLAIN SELECT *, count(*) over (jz) as jesse, count(*) over (oa) as omer FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
                                          QUERY PLAN
    ---------------------------------------------------------------------------------------
     Gather Motion 3:1  (slice1; segments: 3)  (cost=7208.12..7987.12 rows=77900 width=12)
       ->  Window  (cost=7208.12..7987.12 rows=25967 width=12)
             Partition By: a
             Order By (level 1): b
             Order By (level 2): c
             ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
                   Sort Key: a, b, c
                   ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
     Settings:  optimizer=off
     Optimizer status: legacy query optimizer
    (10 rows)

    ```

Renames aside, we do see an extra sort after your change. And depending on the data this could be quite taxing on performance.

hlinnaka commented:
Thanks!

On 09/19/2017 04:39 AM, Jesse Zhang wrote:
> I believe this change has resulted in inferior plans. Haven't looked too closely at exactly what's causing it, but here are the steps to reproduce:
> ...
> 1. Plans after your patch:
> 
>      ```SQL
>      EXPLAIN SELECT *, count(*) over (oa) as omer, count(*) over (jz) as jesse FROM foo WINDOW jz AS (PARTITION BY a ORDER BY b), oa AS (PARTITION BY a ORDER BY b, c);
>                                                QUERY PLAN
>      ----------------------------------------------------------------------------------------------
>       Gather Motion 3:1  (slice1; segments: 3)  (cost=14705.73..16068.98 rows=77900 width=12)
>         ->  Subquery Scan coplan  (cost=14705.73..16068.98 rows=25967 width=12)
>               ->  WindowAgg  (cost=14705.73..15289.98 rows=25967 width=12)
>                     Partition By: foo.a
>                     Order By: foo.b, foo.c
>                     ->  Sort  (cost=14705.73..14900.48 rows=25967 width=12)
>                           Sort Key: foo.a, foo.b, foo.c
>                           ->  WindowAgg  (cost=7208.12..7597.62 rows=25967 width=12)
>                                 Partition By: foo.a
>                                 Order By: foo.b
>                                 ->  Sort  (cost=7208.12..7402.87 rows=25967 width=12)
>                                       Sort Key: foo.a, foo.b
>                                       ->  Seq Scan on foo  (cost=0.00..879.00 rows=25967 width=12)
>       Settings:  optimizer=off
>       Optimizer status: legacy query optimizer
>      (15 rows)
> ...
> Renames aside, we do see an extra sort after your change. And
> depending on the data this could be quite taxing on performance.

In a nutshell, the planner code sorts all the Windows, before it creates 
the WindowAgg nodes that are stacked on top of each other. Before this 
PR, the WindowAggs were collapsed to one, with different "levels", and 
it avoided the sorts. But now we create separate WindowAggs, it depends 
on the order they're stacked, whether you get one or two Sorts.

Neither ORCA nor PostgreSQL do anything smart about that, so it depends 
on the order that you happen to use the Windows in the query, whether 
you get one or two sorts. But the GPDB planner now actually pessimizes 
that case. It collects all the Windows in an array, and sorts it so that 
the Window with the shortest ORDER BY pathkeys goes to the bottom. So 
you get the worst possible case, no matter which order you specify the 
windows in the query.

I added a commit to this PR to address that. I tried to change the code 
that sorts the windows, so that the longest keys would go to the bottom, 
but I could not get that to work for some reason. So I changed tack, and 
instead added some code to where we add the Sort nodes, to "peek ahead" 
at the next Window to see if the ORDER BY of the next window is the same 
as the current one's, but with some extra columns, and uses that one if so.

So that's been fixed now, I believe, although I'm not 100% happy with 
the solution.

- Heikki

d commented:
I don't think you updated the tests to reflect the plan change

hlinnaka commented:
On 09/19/2017 07:04 PM, Jesse Zhang wrote:
> I don't think you updated the tests to reflect the plan change

It was actually bug in the code - the plan in the expected output was 
correct, but the code didn't produce that plan. I must've forgotten to 
"git add" or something earlier, because clearly I got the plan that was 
memorized in the expected output at some point.

Anyway, fixed now. I expect all the regression tests to pass now, but 
fingers crossed. There was apparently a crash in the icw_orca test in 
the previous PR pipeline run, which I don't think I saw earlier, so I 
believe it was a fluke and will go away when on next run, but we'll see.

- Heikki

pivotal-mike commented:
@schubert for visibility

hlinnaka commented:
Pushed so that we can move forward. But QP folk, if you still have any comments, please speak up.
